packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1512,6 +1519,58 @@ export class ContainerRuntime
 	 * The options to apply to this ContainerRuntime instance (including internal options hidden from the public API)
 	 */
 	private readonly runtimeOptions: Readonly<Required<IContainerRuntimeOptionsInternal>>;
+	public readonly pkgVersion = pkgVersion;
+	/**
+	 * The current generation of the Runtime layer. This is used to ensure compatibility between the Runtime and
+	 * other layers.
+	 */
+	private readonly generation = currentRuntimeGeneration;
+	/**
+	 * A list of features that the Loader layer is required to support to be compatible with this Runtime.
+	 */
+	private readonly requiredFeaturesFromLoader: string[] = [];
+	/**
+	 * A list of features supported by the Runtime layer. This is exposed to the Loader layer which uses
+	 * it to determine if the Runtime is compatible with the Loader.
+	 */
+	private readonly supportedFeaturesForLoader: ReadonlyMap<string, unknown> = new Map([
+		/* This is the minimum generation of the Loader this Runtime supports. */
+		["minSupportedGeneration", 1],
+	]);
+	public get supportedFeatures(): ReadonlyMap<string, unknown> {
+		return this.supportedFeaturesForLoader;
+	}
+	private validateLoaderCompatibility(
+		loaderSupportedFeatures?: ReadonlyMap<string, unknown>,
+		loaderVersion?: string,
+	): void {
+		if (loaderSupportedFeatures === undefined) {
+			return;
+		}
+
+		const result = checkLayerCompatibility(
+			this.requiredFeaturesFromLoader,
+			this.generation,
+			loaderSupportedFeatures,
+		);
+		if (!result.compatible) {
+			const error = new UsageError("Runtime is not compatible with Loader", {
+				runtimeVersion: pkgVersion,
+				loaderVersion,
+				runtimeGeneration: this.generation,
+				minSupportedGeneration: loaderSupportedFeatures.get(
+					"minSupportedGeneration",
+				) as number,
+				isGenerationCompatible: result.generationCompatible,
+				unsupportedFeatures:
+					result.unsupportedFeatures.length === 0
+						? undefined
+						: JSON.stringify(result.unsupportedFeatures),
+			});
+			this.closeFn(error);
+			throw error;
```

Comment:
Question for my understanding: Do we need to throw the error here if we are already closing with an error? Or, where is the error caught?

Asking because I have seen host apps receive unhandled errors, and I wonder if this will result in one of them.

Reply:
This is a great question.  I think we are over-eager in close-then-throw.  In most cases we can simply throw and higher up code of ours will close.  It's safer this way, but kind of paranoid.

Ties into the bigger discussion of whether we over-use throwing and should be using "result" types that return a success or failure object with strongly typed error info.  Out of scope for this PR obviously :P